// saucelabs-finder/lib/index.js
// Generated by CoffeeScript 1.9.3
var Finder, _,
  slice = [].slice;

_ = require('lodash');

Finder = (function() {
  function Finder() {}

  Finder.prototype.alias = {
    ie: 'internet explorer'
  };

  Finder.prototype.load = function(browsers) {
    var base, browser, i, latest, len, oldest, ref, ref1, ref2, results, type, versions;
    this.browsers = _.sortBy(browsers, function(obj) {
      return parseFloat(obj.short_version);
    });
    this.types = {};
    ref = this.browsers;
    for (i = 0, len = ref.length; i < len; i++) {
      browser = ref[i];
      type = (ref1 = browser.device) != null ? ref1 : browser.api_name;
      if ((base = this.types)[type] == null) {
        base[type] = [];
      }
      this.types[type].push(browser);
    }
    this.index = {};
    ref2 = this.types;
    results = [];
    for (type in ref2) {
      browsers = ref2[type];
      versions = _.uniq(_.pluck(browsers, 'short_version'));
      oldest = _.min(versions);
      latest = _.max(versions);
      results.push(this.index[type] = {
        versions: versions,
        oldest: oldest,
        latest: latest
      });
    }
    return results;
  };

  Finder.prototype.find = function(type, versions, os) {
    var api_name, begin, browser, end, found, i, j, len, len1, name1, ref, ref1, short_version, suitable, unsuitable, version;
    if (versions == null) {
      versions = 'latest';
    }
    if (os == null) {
      os = null;
    }
    if (this.alias[type] != null) {
      type = this.alias[type];
    }
    found = {};
    if (this.index[type] == null) {
      return found;
    }
    if (typeof versions === 'string') {
      versions = versions.split(',');
    }
    for (i = 0, len = versions.length; i < len; i++) {
      version = versions[i];
      ref = new String(version).split('..'), begin = ref[0], end = ref[1];
      if (begin === 'oldest' || begin === '') {
        begin = this.index[type].oldest;
      }
      if (begin === 'latest') {
        begin = this.index[type].latest;
      }
      if (end === 'latest' || end === '') {
        end = this.index[type].latest;
      }
      ref1 = this.types[type];
      for (j = 0, len1 = ref1.length; j < len1; j++) {
        browser = ref1[j];
        api_name = browser.api_name, short_version = browser.short_version;
        suitable = end != null ? (parseFloat(begin) <= short_version && short_version <= parseFloat(end)) : short_version === begin;
        if (!suitable) {
          continue;
        }
        unsuitable = (os != null) && browser.os.indexOf(os) !== 0;
        if (unsuitable) {
          continue;
        }
        if (found[name1 = api_name + '@' + short_version] == null) {
          found[name1] = browser;
        }
      }
    }
    return found;
  };

  Finder.prototype.available = function(all) {
    var available, browsers, i, len, machines, name, os, ref, ref1, type, version;
    if (all == null) {
      all = false;
    }
    available = {};
    ref = ['chrome', 'internet explorer', 'firefox', 'safari', 'iphone', 'ipad', 'android'];
    for (i = 0, len = ref.length; i < len; i++) {
      type = ref[i];
      available[type] = {};
    }
    ref1 = this.types;
    for (type in ref1) {
      browsers = ref1[type];
      if (!(all || (available[type] != null))) {
        continue;
      }
      machines = _.uniq((function() {
        var j, len1, ref2, ref3, results;
        ref2 = _.uniq(_.pluck(browsers, 'os'));
        results = [];
        for (j = 0, len1 = ref2.length; j < len1; j++) {
          os = ref2[j];
          ref3 = os.split(/\s+/), name = ref3[0], version = 2 <= ref3.length ? slice.call(ref3, 1) : [];
          results.push(name);
        }
        return results;
      })());
      if (available[type] == null) {
        available[type] = {};
      }
      available[type].versions = _.sortBy(_.uniq(_.pluck(browsers, 'short_version')));
      available[type].platforms = _.sortBy(machines);
    }
    return available;
  };

  return Finder;

})();

module.exports = new Finder;

module.exports.Finder = Finder;
