// karma-saucelabs-launcher/lib/index.js
// Generated by CoffeeScript 1.10.0
var Launcher, Promise, Reporter, SauceLabs, Webdriver, YAML, connectOptions, finder, path, ref, ref1, sauceConnect;

Webdriver = require('./webdriver');

Promise = require('bluebird');

sauceConnect = require('sauce-connect-launcher');

SauceLabs = require('saucelabs');

YAML = require('yamljs');

path = require('path');

finder = require('saucelabs-finder');

connectOptions = {
  username: (ref = process.env.SAUCE_USERNAME) != null ? ref : 'SAUCE_USERNAME',
  accessKey: (ref1 = process.env.SAUCE_ACCESS_KEY) != null ? ref1 : 'SAUCE_ACCESS_KEY',
  tunnelIdentifier: 'launcher:sauce' + Date.now()
};

Launcher = (function() {
  function Launcher(logger, baseLauncherDecorator, config, emitter, sessions) {
    var exitCode, log, sauceConnectLauncher, webdriver;
    log = logger.create('launcher:sauce');
    baseLauncherDecorator(this);
    emitter.setMaxListeners(0);
    config.browserNoActivityTimeout = 0;
    this.name = 'SauceLauncher';
    this._retryLimit = -1;
    this.error = 'failure';
    exitCode = 1;
    sauceConnectLauncher = null;
    webdriver = null;
    this.on('start', function(uri) {
      return new Promise(function(resolve, reject) {
        var saucelabs;
        log.debug('Parse .saucelabs.yml');
        saucelabs = new SauceLabs;
        return saucelabs.getWebDriverBrowsers(function(error, browsers) {
          var browser, i, len, name, platform, ref2, ref3, type, url, version;
          if (error != null) {
            return reject(error);
          }
          finder.load(browsers);
          finder.yaml = YAML.load(process.cwd() + path.sep + '.saucelabs.yml');
          ref2 = finder.yaml.browsers;
          for (i = 0, len = ref2.length; i < len; i++) {
            type = ref2[i];
            name = type.name, version = type.version, platform = type.platform;
            ref3 = finder.find(name, version, platform);
            for (version in ref3) {
              browser = ref3[version];
              if (version === 'internet explorer@6') {
                log.warn('internet explorer@6 is unsupported');
                continue;
              }
              log.debug(JSON.stringify(browser));
              sessions.push(browser);
            }
          }
          url = 'https://saucelabs.com/rest/v1/info/browsers/webdriver';
          log.info('Found %s wds in %s', sessions.length, url);
          return resolve();
        });
      }).then(function() {
        return new Promise(function(resolve, reject) {
          var options;
          options = connectOptions;
          options.logger = log.debug.bind(log);
          return sauceConnect(options, function(error, connect) {
            if (error != null) {
              return reject(error);
            }
            sauceConnectLauncher = connect;
            return resolve(null);
          });
        });
      }).then(function() {
        if (sessions.length === 0) {
          return Promise.resolve(1);
        }
        webdriver = new Webdriver(uri, sessions, {
          tunnelIdentifier: connectOptions.tunnelIdentifier,
          username: connectOptions.username,
          accessKey: connectOptions.accessKey,
          logger: logger
        });
        process.nextTick(function() {
          var i, id, len, ref2, results, session;
          emitter.on('sauce:onBrowserComplete', function(id) {
            return webdriver.complete(id);
          });
          emitter.on('sauce:onBrowserError', function(id) {
            return webdriver.complete(id);
          });
          ref2 = webdriver.sessions;
          results = [];
          for (id = i = 0, len = ref2.length; i < len; id = ++i) {
            session = ref2[id];
            results.push(webdriver.boot(id));
          }
          return results;
        });
        return webdriver.done.promise;
      }).spread(function(exitCode, passed, failed, count, msec) {
        log.info('%d passed, %d failed. Total %d browsers(%s sec).', passed, failed, count, msec / 1000);
        return emitter.emit('run_complete', [], {
          exitCode: exitCode
        });
      });
    });
    this.once('kill', function(done) {
      var i, len, queues, ref2, ref3, session;
      queues = [];
      if (sauceConnectLauncher) {
        queues.push(new Promise(function(resolve) {
          sauceConnectLauncher.close(resolve);
          return sauceConnectLauncher = null;
        }));
      } else {
        queues.push(Promise.resolve());
      }
      if (webdriver != null ? webdriver.sessions.length : void 0) {
        ref2 = webdriver.sessions;
        for (i = 0, len = ref2.length; i < len; i++) {
          session = ref2[i];
          if ((ref3 = session.driver) != null ? ref3.passed : void 0) {
            queues.push(session.driver.passed(false));
          }
        }
      }
      return Promise.all(queues)["finally"](function() {
        return done();
      });
    });
  }

  return Launcher;

})();

Reporter = require('./reporter');

module.exports = {
  'sessions': ['value', []],
  'launcher:Sauce': ['type', Launcher],
  'reporter:sauce': ['type', Reporter]
};
