// karma-saucelabs-launcher/lib/reporter.js
// Generated by CoffeeScript 1.10.0
var Reporter, requrest,
  slice = [].slice;

requrest = require('request');

Reporter = (function() {
  function Reporter(baseReporterDecorator, capturedBrowsers, emitter, sessions) {
    baseReporterDecorator(this);
    this.TOTAL_SUCCESS = '';
    this.TOTAL_FAILED = '';
    this.onBrowserStart = function(browser) {
      emitter.emit('sauce:onBrowserStart', browser);
      browser.reconnect = function() {};
      browser.onDisconnect = function() {};
      return capturedBrowsers.remove(browser);
    };
    this.onBrowserComplete = function(browser) {
      var session;
      session = sessions[browser.id];
      session.lastResult = browser.lastResult;
      return emitter.emit('sauce:onBrowserComplete', browser.id);
    };
    this.specFailure = function(browser, result) {
      var log, session;
      session = sessions[browser.id];
      log = session.driver.log;
      return log.error('\nFailure: %s %s\n%s', result.suite.join(' '), result.description, result.log);
    };
    this.onBrowserError = function(browser, error) {
      var i, j, line, log, ref, ref1, ref2, ref3, ref4, ref5, schemas, session, uri, url;
      session = sessions[browser.id];
      session.lastResult = browser.lastResult;
      log = session.driver.log;
      url = (ref = (ref1 = error.match(/http:.+/)) != null ? ref1[0] : void 0) != null ? ref : '';
      ref2 = url.split(':'), schemas = 2 <= ref2.length ? slice.call(ref2, 0, i = ref2.length - 1) : (i = 0, []), line = ref2[i++];
      uri = schemas.join(':');
      if (!(line >= 0)) {
        log.error('\n%s', error);
        emitter.emit('sauce:onBrowserError', browser.id);
        return;
      }
      url = (ref3 = (ref4 = error.match(/http:.+/)) != null ? ref4[0] : void 0) != null ? ref3 : '';
      ref5 = url.split(':'), schemas = 2 <= ref5.length ? slice.call(ref5, 0, j = ref5.length - 1) : (j = 0, []), line = ref5[j++];
      uri = schemas.join(':');
      return requrest(uri, function(ignoredError, response) {
        var codeChunk, ref6;
        codeChunk = response != null ? (ref6 = response.body) != null ? ref6.split('\n')[line - 1] : void 0 : void 0;
        if (codeChunk) {
          log.error('\n%s\n%s: %s', error, line, codeChunk);
        } else {
          log.error('\n%s', error);
        }
        return emitter.emit('sauce:onBrowserError', browser.id);
      });
    };
  }

  return Reporter;

})();

module.exports = Reporter;
