// karma-saucelabs-launcher/lib/webdriver.js
// Generated by CoffeeScript 1.10.0
var Promise, Webdriver, error1, path, pkg, wd,
  bind = function(fn, me){ return function(){ return fn.apply(me, arguments); }; };

Promise = require('bluebird');

wd = require('wd');

try {
  path = require('path');
  pkg = require(process.cwd() + path.sep + 'package');
} catch (error1) {
  pkg = {
    name: 'Karma test'
  };
}

Webdriver = (function() {
  function Webdriver(uri, sessions, options) {
    var qs, ref, ref1, ref2, ref3;
    this.sessions = sessions != null ? sessions : [];
    if (options == null) {
      options = {};
    }
    this.boot = bind(this.boot, this);
    ref = uri.split('?'), this.url = ref[0], qs = ref[1];
    this.tunnelIdentifier = options.tunnelIdentifier;
    this.username = (ref1 = options.username) != null ? ref1 : 'SAUCE_USERNAME';
    this.accessKey = (ref2 = options.accessKey) != null ? ref2 : 'SAUCE_ACCESS_KEY';
    this.begin = Date.now();
    this.current = 0;
    this.concurrency = (ref3 = options.concurrency) != null ? ref3 : 5;
    this.active = 0;
    this.passed = 0;
    this.failed = 0;
    this.logger = options.logger;
    this.log = this.createLogger('wd');
    this.done = Promise.defer();
  }

  Webdriver.prototype.createLogger = function(name) {
    if (this.logger) {
      return this.logger.create(name);
    } else {
      return {
        debug: function() {},
        info: function() {},
        warn: function() {},
        error: function() {}
      };
    }
  };

  Webdriver.prototype.queue = function(browser) {
    var api_name, os, short_version;
    api_name = browser.api_name, short_version = browser.short_version, os = browser.os;
    return this.log.info('Queuing %s@%s at %s', api_name, short_version, os);
  };

  Webdriver.prototype.boot = function(id) {
    var api_name, browser, options, os, short_version;
    browser = this.sessions[id];
    if (this.active >= this.concurrency) {
      return this.queue(browser);
    }
    if (this.current === this.sessions.length && this.active === 0) {
      return this.finish();
    }
    if (browser == null) {
      return;
    }
    this.current++;
    this.active++;
    api_name = browser.api_name, short_version = browser.short_version, os = browser.os;
    this.log.debug('Progress %s/%s Concurrency %s/%s Last %s', this.current, this.sessions.length, this.active, this.concurrency, this.current === this.sessions.length && this.active === 0);
    this.log.info('Start %s@%s at %s', api_name, short_version, os);
    options = {};
    options.browserName = browser.api_name;
    options.version = browser.short_version;
    options.platform = browser.os;
    options.tags = [];
    options.name = pkg.name;
    options['tunnel-identifier'] = this.tunnelIdentifier;
    options['record-video'] = true;
    options['record-screenshots'] = true;
    options['device-orientation'] = null;
    options['disable-popup-handler'] = true;
    options['build'] = process.env.TRAVIS_BUILD_NUMBER;
    if (options['build'] == null) {
      options['build'] = process.env.BUILD_NUMBER;
    }
    if (options['build'] == null) {
      options['build'] = process.env.BUILD_TAG;
    }
    if (options['build'] == null) {
      options['build'] = process.env.CIRCLE_BUILD_NUM;
    }
    return this.sessions[id].driver = this.launch(id, options);
  };

  Webdriver.prototype.launch = function(id, options) {
    var args, browserName, driver, heartbeat, heartbeatFail, log, platform, uri, version;
    args = ['ondemand.saucelabs.com', 80, this.username, this.accessKey];
    browserName = options.browserName, version = options.version, platform = options.platform;
    log = this.createLogger("wd:" + browserName + "@" + version + "(" + platform + ")");
    driver = wd.promiseChainRemote.apply(wd, args);
    driver.on('status', function(info) {
      return log.debug(info);
    });
    driver.on('command', function(eventType, command, response) {
      return log.debug(eventType, command, response || '');
    });
    driver.on('http', function(meth, path, data) {
      return log.debug(meth, path, data || '');
    });
    uri = this.url + '?id=' + id;
    driver.init(options).get(uri);
    heartbeatFail = 0;
    heartbeat = setInterval((function(_this) {
      return function() {
        log.debug('Heartbeat(%s) in %s@%s at %s', heartbeatFail, browserName, version, platform);
        return driver.title().then(null, function(error) {
          heartbeatFail++;
          if (heartbeatFail >= 2) {
            _this.complete(id);
          }
          return log.warn('Heartbeat(%s) %s', heartbeatFail, JSON.stringify(error));
        });
      };
    })(this), 60000);
    driver.log = log;
    driver.clearHeartbeat = function() {
      clearInterval(heartbeat);
      return log.debug('Stop heartbeat in %s@%s at %s', browserName, version, platform);
    };
    driver.passed = (function(_this) {
      return function(pass) {
        return new Promise(function(resolve) {
          var quitId, quitted;
          quitId = setTimeout((function() {
            return quitted();
          }), 2000);
          quitted = function() {
            if (!quitId) {
              return;
            }
            quitId = null;
            _this.active--;
            _this.boot(_this.current);
            return resolve();
          };
          return driver.sauceJobStatus(pass).get('about:blank').quit(quitted);
        });
      };
    })(this);
    return driver;
  };

  Webdriver.prototype.complete = function(id) {
    var api_name, driver, lastResult, os, pass, session, short_version;
    session = this.sessions[id];
    if (!(session != null ? session.driver : void 0)) {
      return this.log.warn('Can not complete invalid session(%s)', id);
    }
    driver = session.driver, lastResult = session.lastResult, api_name = session.api_name, short_version = session.short_version, os = session.os;
    if (driver.clearHeartbeat) {
      driver.clearHeartbeat();
    }
    if (lastResult == null) {
      lastResult = {
        crash: true
      };
    }
    pass = !(lastResult.crash || lastResult.failed || lastResult.error);
    if (pass) {
      this.passed++;
      this.log.info('Passed %s@%s at %s (passed %s/ total %s)', api_name, short_version, os, lastResult.success, lastResult.total);
    } else {
      this.failed++;
      this.log.error('Failed %s@%s at %s (passed %s/ failed %s/ total %s)', api_name, short_version, os, lastResult.success | 0, lastResult.failed | 0, lastResult.total | 0);
    }
    this.log.debug('Shutting down %s@%s at %s', api_name, short_version, os);
    driver.passed(pass);
    return delete session.driver;
  };

  Webdriver.prototype.finish = function() {
    var code;
    code = this.passed === this.sessions.length ? 0 : 1;
    this.log.debug('Exit code is %d', code);
    return this.done.resolve([code, this.passed, this.failed, this.sessions.length, Date.now() - this.begin]);
  };

  return Webdriver;

})();

module.exports = Webdriver;
